File: sandbox-python/brimble_sandbox/resources/sandbox_handle.py
"""Sandbox handle returned from create()/get() with direct runtime operations."""

from __future__ import annotations

import time
from collections.abc import Callable, Mapping
from typing import TYPE_CHECKING, cast

from ..streaming import ByteStream, ExecStream

from ..constants import DEFAULT_SANDBOX_READY_POLL_INTERVAL_MS, DEFAULT_SANDBOX_READY_TIMEOUT_MS
from ..enums import SandboxStatus
from ..transport import RetryOptions
from ..types import (
    BatchFileUploadFileInput,
    CodeInput,
    CreateSnapshotInput,
    ExecInput,
    FileUploadBody,
    Pagination,
    StatsQuery,
    WaitPreference,
)
from ..types import ExecResult
from ..types import Paginated, Snapshot
from ..types import SandboxBatchFileUpload

if TYPE_CHECKING:
    from .sandboxes import SandboxesResource


class SandboxHandle:
    """Convenience handle for one sandbox returned by `sandboxes.create(...)` or `sandboxes.get(...)`."""

    def __init__(self, sandboxes: "SandboxesResource", state: Mapping[str, object]) -> None:
        self._sandboxes = sandboxes
        self._state = dict(state)
        self._scope = sandboxes.use(self.id)
        self.snapshots = _SandboxHandleSnapshots(self)

    @property
    def id(self) -> str:
        """Current sandbox id."""
        return str(self._state["id"])

    @property
    def status(self) -> str:
        """Current cached sandbox status."""
        return str(self._state["status"])

    @property
    def data(self) -> dict[str, object]:
        """Current cached sandbox payload."""
        return self._state

    def refresh(
        self,
        *,
        timeout_ms: int | None = None,
        retry: RetryOptions | bool | None = None,
    ) -> dict[str, object]:
        """Refresh sandbox details from the API and update local state."""
        sandbox = self._sandboxes.get_data(self.id, timeout_ms=timeout_ms, retry=retry)
        self._state = dict(sandbox)
        return self._state

    def destroy(
        self,
        *,
        timeout_ms: int | None = None,
        idempotency_key: str | None = None,
        retry: RetryOptions | bool | None = None,
    ) -> None:
        """Destroy this sandbox (idempotent)."""
        self._sandboxes.destroy(self.id, timeout_ms=timeout_ms, idempotency_key=idempotency_key, retry=retry)

    def pause(
        self,
        *,
        timeout_ms: int | None = None,
        idempotency_key: str | None = None,
        retry: RetryOptions | bool | None = None,
    ) -> dict[str, object] | None:
        """Request pause for this sandbox and refresh local state."""
        response = self._sandboxes.pause(self.id, timeout_ms=timeout_ms, idempotency_key=idempotency_key, retry=retry)
        self.refresh(timeout_ms=timeout_ms, retry=retry)
        return response

    def resume(
        self,
        *,
        timeout_ms: int | None = None,
        idempotency_key: str | None = None,
        retry: RetryOptions | bool | None = None,
    ) -> dict[str, object] | None:
        """Request resume for this sandbox and refresh local state."""
        response = self._sandboxes.resume(self.id, timeout_ms=timeout_ms, idempotency_key=idempotency_key, retry=retry)
        self.refresh(timeout_ms=timeout_ms, retry=retry)
        return response

    def wait_until_ready(
        self,
        *,
        timeout_ms: int | None = DEFAULT_SANDBOX_READY_TIMEOUT_MS,
        poll_interval_ms: int | None = DEFAULT_SANDBOX_READY_POLL_INTERVAL_MS,
    ) -> dict[str, object]:
        """Poll until sandbox status becomes `ready`, or raise on timeout."""
        effective_timeout_ms = timeout_ms if timeout_ms is not None else DEFAULT_SANDBOX_READY_TIMEOUT_MS
        effective_poll_interval_ms = (
            poll_interval_ms if poll_interval_ms is not None else DEFAULT_SANDBOX_READY_POLL_INTERVAL_MS
        )

        deadline = time.monotonic() + (effective_timeout_ms / 1000)

        while True:
            sandbox = self.refresh(timeout_ms=effective_timeout_ms)
            if sandbox.get("status") == SandboxStatus.READY:
                return sandbox

            if time.monotonic() >= deadline:
                raise TimeoutError(f"Sandbox {self.id} did not become ready within {effective_timeout_ms}ms")

            time.sleep(max(effective_poll_interval_ms, 1) / 1000)

    def exec(
        self,
        input: ExecInput,
        *,
        timeout_ms: int | None = None,
        retry: RetryOptions | bool | None = None,
        wait_until_ready: WaitPreference | None = None,
        on_stdout: Callable[[str], None] | None = None,
        on_stderr: Callable[[str], None] | None = None,
    ) -> ExecResult | ExecStream:
        """Run a shell command in this sandbox; optionally auto-wait for readiness."""
        self._ensure_ready(wait_until_ready)
        return self._scope.exec(
            input,
            timeout_ms=timeout_ms,
            retry=retry,
            on_stdout=on_stdout,
            on_stderr=on_stderr,
        )

    def run_code(
        self,
        input: CodeInput,
        *,
        timeout_ms: int | None = None,
        retry: RetryOptions | bool | None = None,
        wait_until_ready: WaitPreference | None = None,
        on_stdout: Callable[[str], None] | None = None,
        on_stderr: Callable[[str], None] | None = None,
    ) -> ExecResult | ExecStream:
        """Run a code snippet in this sandbox; optionally auto-wait for readiness."""
        self._ensure_ready(wait_until_ready)
        return self._scope.run_code(
            input,
            timeout_ms=timeout_ms,
            retry=retry,
            on_stdout=on_stdout,
            on_stderr=on_stderr,
        )

    def exec_stream(
        self,
        input: ExecInput,
        *,
        timeout_ms: int | None = None,
        retry: RetryOptions | bool | None = None,
        wait_until_ready: WaitPreference | None = None,
    ) -> ExecStream:
        """Run a shell command and stream parsed stdout/stderr output."""
        self._ensure_ready(wait_until_ready)
        return self._scope.exec_stream(input, timeout_ms=timeout_ms, retry=retry)

    def run_code_stream(
        self,
        input: CodeInput,
        *,
        timeout_ms: int | None = None,
        retry: RetryOptions | bool | None = None,
        wait_until_ready: WaitPreference | None = None,
    ) -> ExecStream:
        """Run a code snippet and stream parsed stdout/stderr output."""
        self._ensure_ready(wait_until_ready)
        return self._scope.run_code_stream(input, timeout_ms=timeout_ms, retry=retry)

    def put_file(
        self,
        path: str,
        body: FileUploadBody,
        *,
        timeout_ms: int | None = None,
        retry: RetryOptions | bool | None = None,
        wait_until_ready: WaitPreference | None = None,
    ) -> None:
        """Upload bytes to this sandbox; optionally auto-wait for readiness."""
        self._ensure_ready(wait_until_ready)
        self._scope.put_file(path, body, timeout_ms=timeout_ms, retry=retry)

    def get_file(
        self,
        path: str,
        *,
        timeout_ms: int | None = None,
        retry: RetryOptions | bool | None = None,
        wait_until_ready: WaitPreference | None = None,
    ) -> ByteStream:
        """Download file bytes from this sandbox; optionally auto-wait for readiness."""
        self._ensure_ready(wait_until_ready)
        return self._scope.get_file(path, timeout_ms=timeout_ms, retry=retry)

    def put_files(
        self,
        files: list[BatchFileUploadFileInput],
        *,
        timeout_ms: int | None = None,
        retry: RetryOptions | bool | None = None,
        wait_until_ready: WaitPreference | None = None,
    ) -> SandboxBatchFileUpload:
        """Upload multiple files to this sandbox; optionally auto-wait for readiness."""
        self._ensure_ready(wait_until_ready)
        return self._scope.put_files(files, timeout_ms=timeout_ms, retry=retry)

    def stats(
        self,
        query: StatsQuery | None = None,
        *,
        timeout_ms: int | None = None,
        retry: RetryOptions | bool | None = None,
        wait_until_ready: WaitPreference | None = None,
    ) -> dict[str, object]:
        """Fetch stats for this sandbox; optionally auto-wait for readiness."""
        self._ensure_ready(wait_until_ready)
        return self._scope.stats(query, timeout_ms=timeout_ms, retry=retry)

    def create_snapshot(
        self,
        input: CreateSnapshotInput,
        *,
        timeout_ms: int | None = None,
        retry: RetryOptions | bool | None = None,
        wait_until_ready: WaitPreference | None = None,
    ) -> Snapshot:
        """Create a snapshot for this sandbox; optionally auto-wait for readiness."""
        self._ensure_ready(wait_until_ready)
        return self._scope.create_snapshot(input, timeout_ms=timeout_ms, retry=retry)

    def list_snapshots(
        self,
        query: Pagination | None = None,
        *,
        timeout_ms: int | None = None,
        retry: RetryOptions | bool | None = None,
        wait_until_ready: WaitPreference | None = None,
    ) -> Paginated:
        """List snapshots for this sandbox; optionally auto-wait for readiness."""
        self._ensure_ready(wait_until_ready)
        return self._scope.list_snapshots(query, timeout_ms=timeout_ms, retry=retry)

    def _ensure_ready(self, wait_until_ready: WaitPreference | None) -> None:
        if self.status == SandboxStatus.READY:
            return

        if wait_until_ready:
            if isinstance(wait_until_ready, dict):
                self.wait_until_ready(
                    timeout_ms=cast(int | None, wait_until_ready.get("timeout_ms")),
                    poll_interval_ms=cast(int | None, wait_until_ready.get("poll_interval_ms")),
                )
                return

            self.wait_until_ready()
            return

        self._assert_ready()

    def _assert_ready(self) -> None:
        if self.status != SandboxStatus.READY:
            raise RuntimeError(
                f"Sandbox {self.id} is {self.status}. Call wait_until_ready() or refresh() before runtime operations."
            )


class _SandboxHandleSnapshots:
    """Friendly snapshot helper namespace on top of SandboxHandle."""

    def __init__(self, handle: SandboxHandle) -> None:
        self._handle = handle

    def create(
        self,
        input: CreateSnapshotInput,
        *,
        timeout_ms: int | None = None,
        retry: RetryOptions | bool | None = None,
        wait_until_ready: WaitPreference | None = None,
    ) -> Snapshot:
        """Create a snapshot for this sandbox."""
        return self._handle.create_snapshot(
            input,
            timeout_ms=timeout_ms,
            retry=retry,
            wait_until_ready=wait_until_ready,
        )

    def list(
        self,
        query: Pagination | None = None,
        *,
        timeout_ms: int | None = None,
        retry: RetryOptions | bool | None = None,
        wait_until_ready: WaitPreference | None = None,
    ) -> Paginated:
        """List snapshots for this sandbox."""
        return self._handle.list_snapshots(
            query,
            timeout_ms=timeout_ms,
            retry=retry,
            wait_until_ready=wait_until_ready,
        )

    def iterate(
        self,
        query: Pagination | None = None,
        *,
        timeout_ms: int | None = None,
        retry: RetryOptions | bool | None = None,
        wait_until_ready: WaitPreference | None = None,
    ) -> list[Snapshot]:
        """Iterate snapshots for this sandbox and return flattened results."""
        page = query["page"] if query and "page" in query else 1
        limit = query["limit"] if query and "limit" in query else 15
        results: list[Snapshot] = []

        while True:
            payload = cast(Pagination, dict(query or {}))
            payload["page"] = page
            payload["limit"] = limit

            data = self.list(
                payload,
                timeout_ms=timeout_ms,
                retry=retry,
                wait_until_ready=wait_until_ready,
            )

            snapshots = cast(list[Snapshot], data.get("data", []))
            results.extend(snapshots)

            total_pages = data["totalPages"]
            if page >= total_pages or len(snapshots) == 0:
                break

            page += 1

        return results

File: sandbox/src/resources/sandbox-handle.ts
import {
  DEFAULT_SANDBOX_LONG_POLL_TIMEOUT_SECONDS,
  DEFAULT_SANDBOX_READY_TIMEOUT_MS,
  DEFAULT_SANDBOX_WAIT_POLL_INTERVAL_MS,
} from '../constants';
import { SandboxStatus } from '../enums';
import { NotFoundError } from '../errors';
import type { RequestOptions } from '../transport/http';
import type {
  AckMessage,
  BatchFileUploadInput,
  BatchFileUploadResponse,
  CodeInput,
  CreateSandboxResult,
  CreateSnapshotInput,
  ExecInput,
  ExecResult,
  FileUploadBody,
  Paginated,
  Pagination,
  Sandbox,
  SandboxRuntimeOptions,
  Snapshot,
  Stats,
  StatsQuery,
  WaitPreference,
  WaitUntilReadyOptions,
  UpdateSandboxEgressInput,
} from '../types';
import { ScopedSandboxResource } from './scoped-sandbox';
import type { SandboxesResource } from './sandboxes';
import type { ExecStream } from '../streaming';

function delay(ms: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}

export class SandboxHandle {
  private readonly sandboxes: SandboxesResource;
  private readonly scope: ScopedSandboxResource;
  private sandboxState: Sandbox | CreateSandboxResult;

  /** Snapshot operations grouped under a dedicated namespace. */
  public readonly snapshots: {
    create: (input: CreateSnapshotInput, options?: RequestOptions) => Promise<Snapshot>;
    list: (query?: Pagination, options?: RequestOptions) => Promise<Paginated<Snapshot>>;
  };

  /** @internal Create a sandbox handle from create/get responses. */
  public constructor(sandboxes: SandboxesResource, state: Sandbox | CreateSandboxResult) {
    this.sandboxes = sandboxes;
    this.sandboxState = state;
    this.scope = this.sandboxes.use(state.id);

    this.snapshots = {
      create: (input, options) => this.createSnapshot(input, options),
      list: (query = {}, options) => this.listSnapshots(query, options),
    };
  }

  /** Current sandbox id. */
  public get id(): string {
    return this.sandboxState.id;
  }

  /** Current cached sandbox status. */
  public get status(): SandboxStatus {
    return this.sandboxState.status;
  }

  /** Current cached sandbox payload. */
  public get data(): Sandbox | CreateSandboxResult {
    return this.sandboxState;
  }

  /** Refresh sandbox details from the API and update local state. */
  public async refresh(options?: RequestOptions): Promise<Sandbox> {
    const sandbox = await this.sandboxes.getData(this.id, options);
    this.sandboxState = sandbox;
    return sandbox;
  }

  /** Destroy this sandbox (idempotent). */
  public async destroy(options?: RequestOptions): Promise<void> {
    await this.sandboxes.destroy(this.id, options);
  }

  /** Request pause for this sandbox and refresh cached state. */
  public async pause(options?: RequestOptions): Promise<AckMessage | undefined> {
    const response = await this.sandboxes.pause(this.id, options);
    await this.refresh(options);
    return response;
  }

  /** Request resume for this sandbox and refresh cached state. */
  public async resume(options?: RequestOptions): Promise<AckMessage | undefined> {
    const response = await this.sandboxes.resume(this.id, options);
    await this.refresh(options);
    return response;
  }

  /** Update outbound network policy for this sandbox and refresh local state. */
  public async updateEgress(input: UpdateSandboxEgressInput, options?: RequestOptions): Promise<Sandbox> {
    const sandbox = await this.sandboxes.updateEgress(this.id, input, options);
    this.sandboxState = sandbox;
    return sandbox;
  }

  /**
   * Wait until the sandbox becomes `ready`.
   * Uses server long-poll when available; falls back to fast polling on 404.
   */
  public async waitUntilReady(options: WaitUntilReadyOptions = {}): Promise<Sandbox> {
    if (this.status === SandboxStatus.Ready) {
      return this.refresh({ signal: options.signal });
    }

    const timeoutMs = options.timeoutMs ?? DEFAULT_SANDBOX_READY_TIMEOUT_MS;
    const deadline = Date.now() + timeoutMs;

    while (true) {
      if (options.signal?.aborted) {
        throw new Error('waitUntilReady aborted by signal');
      }

      const remainingMs = deadline - Date.now();
      if (remainingMs <= 0) {
        throw new Error(`Sandbox ${this.id} did not become ready within ${timeoutMs}ms`);
      }

      const waitSeconds = Math.min(
        DEFAULT_SANDBOX_LONG_POLL_TIMEOUT_SECONDS,
        Math.max(1, Math.ceil(remainingMs / 1000)),
      );

      try {
        const sandbox = await this.sandboxes.waitData(
          this.id,
          { timeoutSeconds: waitSeconds, status: SandboxStatus.Ready },
          { signal: options.signal },
        );
        this.sandboxState = sandbox;
        if (sandbox.status === SandboxStatus.Ready) {
          return sandbox;
        }
        if (sandbox.status === SandboxStatus.Failed) {
          throw new Error(`Sandbox ${this.id} failed to provision`);
        }
      } catch (error) {
        if (error instanceof NotFoundError) {
          return this.pollUntilReady(options, deadline);
        }
        throw error;
      }
    }
  }

  private async pollUntilReady(options: WaitUntilReadyOptions, deadline: number): Promise<Sandbox> {
    const pollIntervalMs = options.pollIntervalMs ?? DEFAULT_SANDBOX_WAIT_POLL_INTERVAL_MS;

    while (true) {
      if (options.signal?.aborted) {
        throw new Error('waitUntilReady aborted by signal');
      }

      const sandbox = await this.refresh({ signal: options.signal });
      if (sandbox.status === SandboxStatus.Ready) {
        return sandbox;
      }
      if (sandbox.status === SandboxStatus.Failed) {
        throw new Error(`Sandbox ${this.id} failed to provision`);
      }

      if (Date.now() >= deadline) {
        const timeoutMs = options.timeoutMs ?? DEFAULT_SANDBOX_READY_TIMEOUT_MS;
        throw new Error(`Sandbox ${this.id} did not become ready within ${timeoutMs}ms`);
      }

      await delay(pollIntervalMs);
    }
  }

  /**
   * Run a shell command in this sandbox.
   * By default this throws when not ready; set `waitUntilReady` to auto-wait.
   */
  public exec(input: ExecInput & { stream: true }, options?: SandboxRuntimeOptions): Promise<ExecStream>;
  public exec(input: ExecInput, options?: SandboxRuntimeOptions): Promise<ExecResult>;
  public async exec(input: ExecInput, options: SandboxRuntimeOptions = {}): Promise<ExecResult | ExecStream> {
    await this.ensureReady(options.waitUntilReady);
    return this.scope.exec(input, options);
  }

  /**
   * Run a code snippet in this sandbox.
   * By default this throws when not ready; set `waitUntilReady` to auto-wait.
   */
  public runCode(input: CodeInput & { stream: true }, options?: SandboxRuntimeOptions): Promise<ExecStream>;
  public runCode(input: CodeInput, options?: SandboxRuntimeOptions): Promise<ExecResult>;
  public async runCode(input: CodeInput, options: SandboxRuntimeOptions = {}): Promise<ExecResult | ExecStream> {
    await this.ensureReady(options.waitUntilReady);
    return this.scope.runCode(input, options);
  }

  /**
   * Upload a file into this sandbox.
   * By default this throws when not ready; set `waitUntilReady` to auto-wait.
   */
  public async putFile(path: string, body: FileUploadBody, options: SandboxRuntimeOptions = {}): Promise<void> {
    await this.ensureReady(options.waitUntilReady);
    await this.scope.putFile(path, body, options);
  }

  /**
   * Download a file from this sandbox.
   * By default this throws when not ready; set `waitUntilReady` to auto-wait.
   */
  public async getFile(path: string, options: SandboxRuntimeOptions = {}): Promise<AsyncIterable<Uint8Array>> {
    await this.ensureReady(options.waitUntilReady);
    return this.scope.getFile(path, options);
  }

  /**
   * Upload multiple files in one call.
   * By default this throws when not ready; set `waitUntilReady` to auto-wait.
   */
  public async putFiles(files: BatchFileUploadInput[], options: SandboxRuntimeOptions = {}): Promise<BatchFileUploadResponse> {
    await this.ensureReady(options.waitUntilReady);
    return this.scope.putFiles(files, options);
  }

  /**
   * Fetch usage stats for this sandbox.
   * By default this throws when not ready; set `waitUntilReady` to auto-wait.
   */
  public async stats(query: StatsQuery = {}, options: SandboxRuntimeOptions = {}): Promise<Stats> {
    await this.ensureReady(options.waitUntilReady);
    return this.scope.stats(query, options);
  }

  /**
   * Create a snapshot for this sandbox.
   * By default this throws when not ready; set `waitUntilReady` to auto-wait.
   */
  public async createSnapshot(input: CreateSnapshotInput, options: SandboxRuntimeOptions = {}): Promise<Snapshot> {
    await this.ensureReady(options.waitUntilReady);
    return this.scope.createSnapshot(input, options);
  }

  /**
   * List snapshots for this sandbox.
   * By default this throws when not ready; set `waitUntilReady` to auto-wait.
   */
  public async listSnapshots(query: Pagination = {}, options: SandboxRuntimeOptions = {}): Promise<Paginated<Snapshot>> {
    await this.ensureReady(options.waitUntilReady);
    return this.scope.listSnapshots(query, options);
  }

  private async ensureReady(waitUntilReady: WaitPreference | undefined): Promise<void> {
    if (this.status === SandboxStatus.Ready) {
      return;
    }

    if (waitUntilReady) {
      if (typeof waitUntilReady === 'object') {
        await this.waitUntilReady(waitUntilReady);
        return;
      }

      await this.waitUntilReady();
      return;
    }

    this.assertReady();
  }

  private assertReady(): void {
    if (this.status !== SandboxStatus.Ready) {
      throw new Error(`Sandbox ${this.id} is ${this.status}. Call waitUntilReady() or refresh() before runtime operations.`);
    }
  }
}
